ompi/mpi/tool/event_get_info.c

```diff
@@ -0,0 +1,140 @@
+        }
+
+        /* Copy name and description */
+        mpit_copy_string (name, name_len, event->event_name);
+        mpit_copy_string (desc, desc_len, event->event_description);
+
+        // num_elements is INOUT
+        //
+        // Can query number of datatypes, returned in num_elements
+        //   if array_of_datatypes or displacements are NULL, just return data_type_count.
+        // Otherwise, if array_of_datatypes or displacements are not NULL, use num_elements
+        //   as maximum datatypes or displacements returned.
+        //
+        // Unless the user passes the NULL pointer for num_elements, 
+        //   the function returns the number of elements required for this event type.
+        //
+        // If the number of elements used by the event type is larger than the value of num_elements 
+        //   provided by the user, the number of datatype handles and displacements returned in the 
+        //   corresponding arrays is truncated to the value of num_elements passed in by the user.
+
+        max_datatypes = 0;
+        if (num_elements) {
+            if (NULL != array_of_datatypes || NULL != array_of_displacements) {
+                if (*num_elements < (int) (event->event_datatype_count))
+                    max_datatypes = *num_elements;
+                else
+                    max_datatypes = event->event_datatype_count;
+            }
+            *num_elements = event->event_datatype_count;
+        }
+
+        if (max_datatypes) {
+            if (array_of_datatypes) {
+                for (int i = 0 ; i < max_datatypes ; i++) {
+                    ompi_datatype_t *ompi_datatype = NULL;
+
+                    for (int j = 0 ; j < OMPI_DATATYPE_MPI_MAX_PREDEFINED ; ++j) {
+                        if (ompi_datatype_basicDatatypes[j]->super.id == event->event_datatypes[i]->id) {
+                            ompi_datatype = (ompi_datatype_t *) ompi_datatype_basicDatatypes[j];
+                            break;
+                        }
+                    }
+
+                    assert (NULL != ompi_datatype);
+
+                    array_of_datatypes[i] = ompi_datatype;
+                }
+            }
+
+            if (array_of_displacements) {
+                for (int i = 0 ; i < max_datatypes ; i++) {
+                    array_of_displacements[i] = (MPI_Aint) event->event_offsets[i];
+                }
+            }
+
+            *num_elements = max_datatypes;
+        }
+
+        if (verbosity) {
+            *verbosity = event->event_verbosity;
+        }
+
+        if (NULL != enumtype) {
+            *enumtype = event->event_enumerator ? (MPI_T_enum) event->event_enumerator : MPI_T_ENUM_NULL;
+        }
+
+        if (NULL != bind) {
+            *bind = event->event_bind;
+        }
+
+        if (NULL != extent) {
+            *extent = event->event_extent;
+        }
+
+        if (NULL != info) {
+            *info = OBJ_NEW(ompi_info_t);
+        }
+    } while (0);
```

Comment:
Is there a reason to put this in a `do { ... } while(0)`?  I've usually seen that kind of pattern inside preprocessor macros.